FateIT Repo Error Test

kevinrr888 opened the pull request:
Closes #4266
This adds a test `testRepoFails()` to `FateIT` in 2.1 which ensures `undo()` is called and in the expected order when an error occurs in a Repo's `call()` or `isReady()` methods.

keith-turner commented on test/src/main/java/org/apache/accumulo/test/fate/zookeeper/FateIT.java:
```diff
@@ -369,6 +429,69 @@ public void testCancelWhileInCall() throws Exception {
+      assertEquals(NEW, getTxStatus(zk, txid));
+      fate.seedTransaction("TestOperationFails", txid,
+          new TestOperationFails(1, ExceptionLocation.IS_READY), true, "Test Op Fails");
+      assertEquals(SUBMITTED, getTxStatus(zk, txid));
```
There is a race condition with this check, if a sleep is added and the test run will probably see it.  The status could change from SUBMITTED at any time as background fate threads start working on the seeded transaction.

```suggestion
     Thread.sleep(1000);
      assertEquals(SUBMITTED, getTxStatus(zk, txid));
```

keith-turner commented on test/src/main/java/org/apache/accumulo/test/fate/zookeeper/FateIT.java:
```diff
@@ -369,6 +429,69 @@ public void testCancelWhileInCall() throws Exception {
+      // Wait for all the undo() calls to complete
+      undoLatch.await();
+      assertEquals(expectedUndoOrder, TestOperationFails.undoOrder);
+      assertEquals(FAILED_IN_PROGRESS, getTxStatus(zk, txid));
```
There is a race condition with this check, could do the following which should not have a race condition and utilizes and verifies more of the Fate API.  To do the following would need to disable auto cleanup, will make another comment about that.

```suggestion
      assertEquals(FAILED, fate.waitForCompletion(txid));
      assertTrue(fate.getException(txid).getMessage().contains("isReady() failed"));
```

keith-turner commented on test/src/main/java/org/apache/accumulo/test/fate/zookeeper/FateIT.java:
```diff
@@ -369,6 +429,69 @@ public void testCancelWhileInCall() throws Exception {
+      txid = fate.startTransaction();
+      assertEquals(NEW, getTxStatus(zk, txid));
+      fate.seedTransaction("TestOperationFails", txid,
+          new TestOperationFails(1, ExceptionLocation.IS_READY), true, "Test Op Fails");
```
Disabling auto cleanup makes waitForCompletion suggestion in another comment work.

```suggestion
          new TestOperationFails(1, ExceptionLocation.IS_READY), false, "Test Op Fails");
```

keith-turner commented on test/src/main/java/org/apache/accumulo/test/fate/zookeeper/FateIT.java:
```diff
@@ -369,6 +429,69 @@ public void testCancelWhileInCall() throws Exception {
+      // Wait for all the undo() calls to complete
+      undoLatch.await();
+      assertEquals(expectedUndoOrder, TestOperationFails.undoOrder);
+      assertEquals(FAILED_IN_PROGRESS, getTxStatus(zk, txid));
```
```suggestion
      assertEquals(FAILED, fate.waitForCompletion(txid));
      assertTrue(fate.getException(txid).getMessage().contains("call() failed"));
```

kevinrr888 commented:
- Removed `SUBMITTED` check in the new test `testRepoFails` (not needed - 
  tested elsewhere)
- Changed tests `testTransactionStatus` and `testCancelWhileInCall` to start
  the transaction runners after the `SUBMITTED` check (the race condition
existed in these tests too)
- Resolved race condition on checking the `FAILED` state of the
  transaction in `testRepoFails`
